### DISCUSSION TITLE
Correct crontab configuration

### DISCUSSION THREAD

**Discussion by mnawrockiht** [YetiForceCompany/YetiForceCRM#14551] [How to] [answered]
Hello, I have some problems with making Mail - Scanner to work properly. Going through Issues I found #14394.  In my server configuration, there is red row for app_data/cron.php: 	www (apache)=No, cron (CLI)=No, Owner=-, Permissions=0644, also same for "cache/logs/cron/"
So I went through my crontab configuration as in [documentation ](https://yetiforce.com/en/knowledge-base/documentation/administrator-documentation/item/enable-cron) and to me it seems that all is set as in article.

- crontab:
*/2 * * * * php  /var/www/html/cron.php > /var/www/html/cache/logs/cron.log 2>&1

- file permissions of /var/www/html/cron/cron.sh:
-rwxr--r-- 1 www-data www-data  734 Apr 14 12:04 cron.sh

- php path in the file /var/www/html/cron/cron.sh:
export USE_PHP=/usr/bin/php7.4

I found no information on how to fix red rows in "Files/Folders" table in server configuration. Please advise on how to fix it.

**Comment by mariuszkrzaczkowski** [comment]
in which file did you set the crontab?
What permissions does it run with?

**Reply by mnawrockiht** [reply]
Default crontab location:
/var/spool/cron/crontabs
-rw------- 1 root crontab  372 Apr 14 11:38 root

**Reply by mariuszkrzaczkowski** [reply]
cannot be root, enter the name of the owner of the CRM file

*/2 * * * * www-data /var/www/html/cron/cron.sh > /var/www/html/cache/logs/cron.log 2>&1

**Reply by mnawrockiht** [reply]
I've managed to get Mail-Scanner to work as I want it to, found my error somewhere else. So it is working ok with the cron configuration described in the first post, with red rows in server configuration.
So I guess further inspection of cron configuration is only for the purpose of clearing red rows. 
I've changed the owner of /var/www/html/cache/logs/cron.log from root to www-data and red row for "cache/logs/cron/" cleared. 
I've made crontab for www-data (crontab -u www-data -e) and in it (as suggested):
*/2 * * * * www-data /var/www/html/cron/cron.sh > /var/www/html/cache/logs/cron.log 2>&1
I've also commented cron job from root crontab, so it wouldn't take action twice. 
Those actions stopped CRM cron jobs from running. Red row is still there.
I've changed content of www-data crontab to:
*/2 * * * * php /var/www/html/cron.php > /var/www/html/cache/logs/cron.log 2>&1
And CRM cron jobs started running again, but red row is still there. 
Having both cron.php (for php) and cron.sh (for www-data) actions makes CRM cron jobs run twice (duplicating same actions), and also doesn't clear red row.

**Reply by mariuszkrzaczkowski** [reply]
Delete the file /var/www/html/cache/logs/cron.log if it shows red you need to fix it,
because 100% CRM will generate errors, but you do not know about them yet.

**Reply by mnawrockiht** [reply]
Changing the ownership of /var/www/html/cache/logs/cron.log fixed red row for "cache/logs/cron/". Only red row in server configuration that I have now is for "app_data/cron.php".

**Reply by mariuszkrzaczkowski** [reply]
yes no, it does not solve the problem.
you just hide it

**Reply by mnawrockiht** [reply]
Ok, I've deleted /var/www/html/cache/logs/cron.log. It changed nothing. 
Another log file was created as soon as the next cron has been done. 
Still red row for "app_data/cron.php".

**Reply by mariuszkrzaczkowski** [reply]
you need to correct the cron call

**Reply by mnawrockiht** [reply]
Ok, how?
currently "crontab -u www-data -l" shows:
*/2 * * * * php /var/www/html/cron.php > /var/www/html/cache/logs/cron.log 2>&1
#*/2 * * * * www-data /var/www/html/cron/cron.sh > /var/www/html/cache/logs/cron.log 2>&1
How it should look like?

**Reply by mariuszkrzaczkowski** [reply]
it depends on how you run or edit the file or crontab -e -u www-data

**Reply by mnawrockiht** [reply]
Ok, I know how to edit crontab :) 
I mean, how to correct the cron call? 
What in current configuration is incorrect? How should the cron call look?

**Reply by mariuszkrzaczkowski** [reply]
it depends on how you run or edit the file or crontab -e

**Reply by mariuszkrzaczkowski** [reply]
try it like this 
*/2 * * * * www-data sh /var/www/cron/cron.sh > /var/www/cache/logs/cron.log 2>&1

**Reply by mnawrockiht** [reply]
Crontab look like this now:
crontab -u www-data -l
*/2 * * * * php /var/www/html/cron.php > /var/www/html/cache/logs/cron.log 2>&1
*/2 * * * * www-data sh /var/www/html/cron/cron.sh > /var/www/html/cache/logs/cron.log 2>&1

It changed nothing in crontab behavior and still show red row for "app_data/cron.php".

**Reply by mariuszkrzaczkowski** [reply]
> */2 * * * * php /var/www/html/cron.php > /var/www/html/cache/logs/cron.log 2>&1

wrong, to be removed

> */2 * * * * www-data sh /var/www/html/cron/cron.sh > /var/www/html/cache/logs/cron.log 2>&1

correct

after changing the crohn configuration, it is required to delete the files:
- app_data/cron.php
- cache/logs/cron.log

and changing the recursive owner to the correct one because often a wrongly set cron messes with the owner and then there is a problem with permissions

**Comment by mnawrockiht** [comment, marked as answer]
Thank you, Mariusz for helping. I've got it correct now - the whole Files/Folders table in server configuration is clear now. 
What got it working (changes in configuration from the first post):

- crontab: 
Changed cronab from root owned to user owned (crontab -u www-data -e).
In crontab there is:
*/2 * * * * sh /var/www/html/cron/cron.sh > /var/www/html/cache/logs/cron.log 2>&1
Note that there is no user column (no 'www-data' between stars and sh), because it is user owned crontab, no need to specify it. 

- file permissions:
I've deleted /var/www/html/cache/logs/cron.log, running crontab creates a new one, with proper ownership. 
I've checked/corrected file permissions for files (should be "-rwxr--r-- 1 www-data www-data"):
/var/www/html/app/Cron.php
/var/www/html/app_data/cron.php
/var/www/html/cron.php

- php path in the file /var/www/html/cron/cron.sh:
export USE_PHP=/usr/bin/php7.4
No changes here.

I've got my vanilla instalation YF running from official container image. Considering container should be a somewhat ready environment for running provided software I find the process to get the correct configuration of cron quite cumbersome. And documentation is not helping at all. YetiTeam, please consider changing documentation on how to set crontab. Currently in English version there is no mention of crontab ownership (even though in Polish version there is!). This is very confusing.

**Reply by mariuszkrzaczkowski** [reply]
Container or do you maybe mean docker? The docker crontab is properly configured, please see below:
https://github.com/YetiForceCompany/YetiForceCRM/blob/7fffbd8575ab99619e4f0cca041fb204cd84c484/tests/setup/crons.conf#L1
https://github.com/YetiForceCompany/YetiForceCRM/blob/7fffbd8575ab99619e4f0cca041fb204cd84c484/Dockerfile#L40
https://github.com/YetiForceCompany/YetiForceCRM/blob/7fffbd8575ab99619e4f0cca041fb204cd84c484/Dockerfile#L59
